[Project1/조재훈] Notion 클로닝 프로젝트

jae-hun-e opened the pull request:
## 📌 과제 설명

배포 링크 : https://effulgent-cuchufli-54c548.netlify.app/

컴포넌트 구조 : 
![아토믹 디자인 패턴-1 2](https://github.com/prgrms-fe-devcourse/FEDC5-5_Project_Notion_VanillaJS/assets/76520477/b7596d6f-21da-48ac-b545-da3a6bfa3896)



### 공통

- [x]  글 단위를 Document라고 합니다. Document는 Document 여러개를 포함가능.
- [x]  화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링
    - Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링합니다.
    - 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.
    - Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.
- [x]  Document Save API를 이용해 지속적으로 서버에 저장
- [x]  History API를 이용해 SPA 형태로 만듭니다.

### 추가구현

- [x]  notion처럼 document header에 현재 경로의 부모들을 모두 담아 놓음
- [x]  title이 변경되면 sidebar와 header에 실시간 반영
- [x]  request요청을 줄이기 위해서 디바운스 1초 설정
- [x]  #, ##, ### 인식이 있는 rich Editer구현

## 👩‍💻 요구 사항과 구현 내용

1. pages, templates, organisms, molecules 로 나누어 구현
2. 클릭 이벤트에 대해서 depth가 깊어 props로 넘겨주는것이 아닌 커스텀이벤트를 이용해서 구현
3. POST는 1초에 한번씩 보낼 수 있도록 디바운스를 적용하여 구현
4. NotionSideBar, DocumentHeader, title의 내용을 동기화하기 위해 커스텀 이벤트 + localStorage를 이용해서 동기화하여 구현
5. NotionSideBar에서 오픈된 상태를 유지하기 위해 클릭시 localstorage에 저장하여 localstorage에 있는 id값의 document들은 오픈된 상태 유지
6. DocumentHeader에서 현재 document의 부모들을 불러오기 위해 documentPath를 구하는 로직 추가

를 구현하였습니다!

## ✅ PR 포인트 & 궁금한 점

### setting

- esLint를 처음 써보다보니 제대로 세팅하였는지 궁금합니다!!

### 디렉토리 구조 (아토믹 디자인 패턴)

- 아토믹 디자인 패턴 중 atom+molecules로  4단계로 구성하여 이용해 보았습니다.
    
    atom을 뺀 이유는 제가 생각했을 때 atom에는 input, button 같은 기능이 없는 ui만 담당하는 컴포넌트들이 담겨져있다고 생각했고, 이번 프로젝트에서는 앞으로 디자인이 자주 변경될 일이 없을 것이라고 생각하여 한가지의 기능을 담은 molecules을 atom + molecules라고 생각하고 구성하였습니다.
    
    ```
      - molecules : 한가지 기능
      - organisms : 하나의 의미
      - templates : layout 담당
      - pages : data 담당
    ```
    
    이렇게를 기준으로 담고자 했지만 구현을 하다보니 templates에도 데이터 가공을 하게되는 등 구조가 뒤엉키데 되었습니다ㅜㅜ 
    
    초반에 생각했던 아토믹 디장인의 장점들을 제대로 사용하지 못한것 같아 어떤 식으로 분리를 하는게 더 좋을 지 피드백 해주시면 감사하겠습니다!! 
    
    **겪은 어려움**
    
    1. 아토믹 디자인을 적용하려다보니 컴포넌트간의depth도 너무 깊어지고 컴포넌트간의 상호작용을 하는데 어려움이 있었습니다. 
    2. 어떤 molecules에서는 한번만 렌더링 하면 변하지 않지만 사용자 입력에 따라 state가 변하는경우 리렌더링해야해서 setState, render 함수를 만들게 되었습니다. 그러다보니 각 레벨들의 역할과 같은 레벨의 컴포넌트들의 통일성을 잃게되었다는 느낌을 받게되었습니다. 

### 컴포넌트

1. 컴포넌트 분리가 잘 되었는지 궁금합니다!!
2. molecules 컴포넌트가 SRP 원칙을 잘 지켰는지 궁금합니다!!
3. 컴포넌트간의 의존성이 낮은지, 응집도가 높은지 궁금합니다!!
4. styleComponents 처럼 사용하기 위해서 styleInJS라는 utils를 이용해서 스타일을 따로 뺐습니다. 제가 생각했을 때 
    1. 한 파일에서 모든 스타일을 관리하며 classname 중복을 피할 수 있다.
    2. 컴포넌트 내에 로직부분과 스타일 부분을 따로 분리한다.
    3. JS문법으로 사용할 수 있다. 
    
    이러한 특징이 있다고 생각하는데 이게 장점인지 단점인지 헷갈립니다.

jae-hun-e committed:
[feat] 노션클로닝을 위한 기본 세팅

1. 기본 뼈대 생성
  1. 디렉토리 구조 잡기
    - components: 컴포넌트
    - pages: 페이지들
    - services : api호출 등 외부 서비스 관련
    - utils : 공통으로 사용되는 유틸리티 함수

2. prettier, esLint 설정

3. documentTree 로직 구현 완료

jae-hun-e committed:
[feat] 컴포넌트 디자인 패턴 적용(아토믹 디자인)

1. 아토믹 디자인 중 atom + molecules = molecules 로 하여 4단계로 나눔
  - 디자인 초안도 내가 짜고, 앞으로 변경일도 거의 없어서 atom으로 나눌 필요성을 못느낌
  - molecules : 한가지 기능
  - organisms : 하나의 의미
  - templates : layout 담당
  - pages : data 담당
    - 각각의 컴포넌트에 담고있는것을 주석으로 표시

2. documentTree를 아코디언 형태로 변경

jae-hun-e committed:
[feat] NotionSideBar 초기 렌더링 구현

1. 각 구조에 맞게 컴포넌트의 역할을 나눔

jae-hun-e committed:
[feat] DocumentDetail.js 구현

1. NotionSideBar에서 각 document를 누르면 해당 정보를 DocumentDetail에서 보여줌

jae-hun-e committed:
[feat] DeleteDocumentButton 생성, DocumentContent 생성

1. DeleteDocumentButton을 누르면 해당 document가 삭제됨
  A->B->C 구조일 때 B를 삭제하면 BE로직 내에서 C를 root로 올림
  하지만 유저입장에서는 A->C가 더 타당하다고 생각해서 고민중...

2. DocumentContent는 input 이 아니라 div에 contenteditable 속성을 부여해서 editer를 만들었는데 커서가 고정되지 않는것과 한글이 자음과 모음이 분리되는 문제가 발생

jae-hun-e committed:
[feat] DocumentHeader 구현

1. DocumentLinkButton으로 구성된 DocumentHeader를 만들어서 현재 위치를 나타내고 부모의 곳으로 바로 올라갈 수 있음

jae-hun-e committed:
[feat] DocumentHeader 구현

1. DocumentLinkButton으로 구성된 DocumentHeader를 만들어서 현재 위치를 나타내고 부모의 곳으로 바로 올라갈 수 있음

jae-hun-e committed:
[feat] NotionSideBar open 상태 유지
[refector] create, delete Document 의존성 제거(동작 바깥에서 주입)

1. 유저가 open을 누르면 localstorage에 오픈된 document id를 저장해놓고 리렌더링 시 open시켜줌
  - click 시 open,
  - delete 시 삭제

2. document create, delete시 동작을 외부에서 주입 식으로 변경

3. delete에서도 삭제하면 root로 통일

jae-hun-e committed:
[fix]

1. richContent 자소분리 에러로 인해 일단 textarea로 에디터 변경

jae-hun-e committed:
[fix] editor 수정

1. textArea에서 contenteditable 속성의div로 바꾸면서 #, ##, ### 를 입력할 시 h1, h2, h3로 태그를 바꿔줌

jae-hun-e committed:
[Refactor] style 속성 분리(css in js) , main -> app 구조 변경

1. 각각의 style에 해당하는 부분을 따로 빼서 각 태그에 부여해줌으로써 css를 사용했을 때 id, class가 겹치는 문제를 없앰
2. css를 따로 분리 하여 css 수정이 필요하면 한 파일 내에서 모두 수정이 가능하다

3. main.js와 App.js가 동일한 역할을 하고 있기 때문에 main을 없애고 App.js로 통일함

jae-hun-e committed:
[Refactor] 홈페이지 갔을 때는 입력 못하게 막기

1. DocumentContent.js와 EditDocument.js에서 isDisabled를 이용해서 입력을 막아 줌

2. NotionTitle.js에서 이벤트 버블링을 막아 sideBar까지 전파를 막음

jae-hun-e committed:
[Refactor] 스타일 추가, 에러 수정

1. 에러1 : 새 글을 작성하고 다른데 갔다가 다시 돌아오면 title이 안바껴있음
  원인 : HomePage.js의 documentPathData가 이전 데이터 기준으로 작동
  해결방법 : route()할때 this.state = newDocuments로 바꿔줌

jae-hun-e committed:
[Refactor] title 낙천적 업데이트, 현재 위치 표시(sidebar, header)

1. 데이터를 전송하기 전에 title이 변경되는것을 client에게 선반영하여 보여준다
  - localstorage에 임시 저정
  - request 보내고 삭제
  - 사용하는 곳들의 depth가 복잡해지므로 customEvent로 만
  - localstorage 데이터를 불러와서 sidebar, header업데이트 시킴

2. 발견한 에러 : 한글 중복 입력이됨.....

jae-hun-e committed:
[fix] sidebar hover시 버튼 안보는 에러

1. 원인 : DocumentObject.js 렌더링시 innerHTML을 지우고 다시 만드는데 그때 +,x 버튼을 다시 안 만들어줬었음
  해결 : this.render 함수안에 버튼 생성 넣어줌

jae-hun-e committed:
[fix] sidebar 빈공간 클릭 이벤트 발생

1. 원인 : 이벤트 델리게이션을 사용하려고 sidebar 전체에 이벤트를 걸어줬었음
  해결 : $notionsidebar 외부의 공간을 클릭 시 작동 안하게 변경

jae-hun-e committed:
[fix] notionSideBar 오타 수정 + 안쓰는 파일 삭제

1. DocumentHeader.js에서 원래 여러가지 기능을 넣고자 했지만 DocumentPath만 넣게 되면서 DocumentPath를 삭제함

jae-hun-e committed:
[deploy] vercel deploy setting

vercel을 이용해서 deploy하기 위해 404에러처리

배포링크 : https://notion-cloneing-git-5-5working-jae-hun-e.vercel.app/

seongminn commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+  const findDocumentDataById = id => {
+    const documentPath = [];
+
+    const findDocument = (documents, id) => {
```
findDocumentById와 findDocument라는 이름이 중복이 된다고 생각해서 findDocument 내부의 작업을 밖으로 꺼내도 괜찮지 않을까 하는 생각이 들었습니다! 이런 점에서 재훈님이 findDocumentById와 findDocument를 분리한 이유가 궁금합니다!

seongminn commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+      notionSideBar.setState(this.state);
+
+      // document 보여주기
+      const { content } = await documentContent;
```
저는 보통 API 호출 함수에 await 키워드를 붙였는데요, 재훈님께서는 API 호출로 받아온 Promise 객체에 await 키워드를 붙인 거 같네요!! 제가 이해한 게 맞을까요? 맞다면 이렇게 작성하신 이유가 궁금합니다!!

seongminn commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+
+  this.route();
+
+  this.optimisticTitle = () => {
```
낙관적 업데이트로 게시물 제목을 변경하셨군요!! 저도 하려다가 나중으로 미뤄뒀는데, 영감 받고 한 번 도전해보겠습니다ㅎㅎ 👍

seongminn reviewed (commented):
과제하느라 고생 많으셨습니다 재훈님!!

lunarmoon7 commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+      documentDetail.setState({ id: null, title: '첫 화면', content: '내용을 채워주세요', documentPath: [] });
+    } else if (pathname.indexOf('/documents/') === 0) {
+      const [, , postId] = pathname.split('/');
+      const documentContent = request(`/documents/${postId}`);
```
`request`에 `await`키워드를 붙이지 않고, `documentCotent`에 await를 작성하신 이유가 궁금합니다!

lunarmoon7 commented on src/utils/editerUtils.js:
```diff
@@ -0,0 +1,10 @@
+export const moveCursorToEnd = target => {
```
커서 옮기기 👍

lunarmoon7 commented on src/style/tagStyles.js:
```diff
@@ -0,0 +1,78 @@
+  },
+};
+
+export default function styleInJS({ $target, styleTagName }) {
```
클래스명의 중복을 피할 수 있다는 점에서 좋은 것 같아요! 😃

justhumanb2ing reviewed (commented):
과제 하시느라 고생하셨습니다 ㅎㅎ
커피챗때 얘기 나눴던, 아토믹 패턴을 적용한게 인상적이었습니다 👍
혹시 이 부분은 저만 그런건지 모르겠는데, 배포한 주소에서 문서를 클릭하고 새로고침하면 에러가 발생합니다..!🥺

jae-hun-e commented on src/components/molecules/DeleteDocumentButton.js:
```diff
@@ -0,0 +1,20 @@
+export default function DeleteDocumentButton({ $target, onDeleteDocument, isHidden }) {
+  this.$addDocumentButton = document.createElement('button');
+  this.$addDocumentButton.style.visibility = `${isHidden ? 'hidden' : 'visible'}`;
+  styleInJS({ $target: this.$addDocumentButton, styleTagName: 'DeleteDocumentButton' });
```
CSS와 JS를 분리하고 싶기도 했고 각 파일마다 className이 겹치는 것도 싫어서 한번 흉내만 내봤습니다ㅎㅎㅎ

jae-hun-e committed:
[fix] Cvercel deploy setting

jae-hun-e committed:
[fix] vercel deploy -> netlify deploy

hyeonjinan096 reviewed (commented):
프로젝트 하시느라 수고 많으셨습니다!! 처음 보는 코드도 있어서 읽는 것만으로 도움이 많이 되네요👍👍

nnahyunn commented:
아토믹 디자인,,,,! 뭔지 알아봐야지 미루기만 하다가 이번에 재훈님 코드보고 멘토님 글도 함께 읽어보면서 이해할 수 있어서 도움이 많이 됐습니다! 노션프로젝트의 특성을 고려해서 atom과 molecule을 하나의 단위로 잡으신 점이 인상적입니다👍🏻👍🏻 고민의 흔적이 느껴지네요 프로젝트 하시느라 고생하셨습니다😁많이 배워갑니다ㅎㅎ

hoilzz commented on src/components/molecules/DocumentLinkButton.js:
```diff
@@ -0,0 +1,33 @@
+
+  this.render = () => {
+    $title.textContent = this.state;
+    const { pathname } = window.location;
+    if (pathname.indexOf('/documents/') === 0) {
+      const [, , postId] = pathname.split('/');
+      if (Number(postId) === documentId) {
+        $title.style.color = 'blue';
+      }
+    }
```
디자인 시스템 아래에서는 사실 이 로직은 외부 파라미터로 판단하는게 적절할 거 같아요 ㅎㅎ
molecule은 SRP를 지키고 이런 컨텍스트를 배제해야 재사용성이 높아지기 떄문입니다 ㅎㅎ

hoilzz commented on src/utils/saveDocumentTitle.js:
```diff
@@ -0,0 +1,15 @@
+
+const SAVE_DOCUMENT_TITLE_KEY = 'SAVE_DOCUMENT_TITLE_KEY';
+
+export const optimisticUpdate = optimisticTitle => {
```
이 함수의 역할은 SAVE_DOCUMENT_TITLE 이벤트에 대한 이벤트 등록으로 보여집니다.
addSaveDocumentTitleEvent 와 같은 함수이름으로 변경하고 파라미터도 함수명으로 보여질 수 있도록 변경하면 좋을 거 같아요 ㅎㅎ

hoilzz commented on src/utils/saveDocumentTitle.js:
```diff
@@ -0,0 +1,15 @@
+  });
+};
+
+export const saveDocumentTitle = (title, id) => {
```
요기도 dispatchSaveDocumentTitleEvent 와 같이 이벤트 dispatch에 대해 명시적으로 작성하면 좋을 것 같습니다.

hoilzz commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+
+  let tempDocumentData;
+
+  this.route = async () => {
```
route라는 네이밍만 보면 이동할 것 같은 느낌인데요
구현 코드를 보면 route이벤트 핸들러로 보입니다. ㅎㅎ 함수명을 수정하면 좋을 것 같아요

hoilzz commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+    }
+  };
+
+  this.route();
```
페이지 첫 로드시 route 내부에 상태 설정해주는 로직을 실행주고 싶은 것 같은데..
initialize와 같이 따로 메서드를 추가하거나 혹은 각 컴포넌트의 초기값을 onRoute내부 로직대로 설정해주면 어떨까요?

hoilzz commented on src/page/HomePage.js:
```diff
@@ -0,0 +1,95 @@
+    notionSideBar.setState(this.state);
+    documentDetail.setState(tempDocumentData);
+  };
+  optimisticUpdate(this.optimisticTitle);
```
optimistic UI update 기법 같군요  ㅎㅎ

hoilzz commented on src/services/api.js:
```diff
@@ -0,0 +1,20 @@
+const API_BASE_URL = 'https://kdt-frontend.programmers.co.kr';
+
+export const request = async (url = '', options = {}) => {
```
추상화가 적절히 잘 이뤄진 것 같습니다 👍

hoilzz replied:
파라미터명은 사실 url보다는 pathname이 더 적절할 거 같아요 ㅎㅎ

hoilzz reviewed (commented):
중간정도 봤어요 ㅎㅎ 저녁에 이어서 코멘트 남길께요!




> atom을 뺀 이유는 제가 생각했을 때 atom에는 input, button 같은 기능이 없는 ui만 담당하는 컴포넌트들이 담겨져있다고 생각했고, 이번 프로젝트에서는 앞으로 디자인이 자주 변경될 일이 없을 것이라고 생각하여 한가지의 기능을 담은 molecules을 atom + molecules라고 생각하고 구성하였습니다.

>  - molecules : 한가지 기능
>  - organisms : 하나의 의미
>  - templates : layout 담당
>  - pages : data 담당
> 이렇게를 기준으로 담고자 했지만 구현을 하다보니 templates에도 데이터 가공을 하게되는 등 구조가 뒤엉키데 되었습니다ㅜㅜ
> 초반에 생각했던 아토믹 디장인의 장점들을 제대로 사용하지 못한것 같아 어떤 식으로 분리를 하는게 더 좋을 지 피드백 해주시면 감사하겠습니다!!

> 겪은 어려움

> 아토믹 디자인을 적용하려다보니 컴포넌트간의depth도 너무 깊어지고 컴포넌트간의 상호작용을 하는데 어려움이 있었습니다.
> 어떤 molecules에서는 한번만 렌더링 하면 변하지 않지만 사용자 입력에 따라 state가 변하는경우 리렌더링해야해서 setState, render 함수를 > 만들게 되었습니다. 그러다보니 각 레벨들의 역할과 같은 레벨의 컴포넌트들의 통일성을 잃게되었다는 느낌을 받게되었습니다.

- 이건 커피챗 때 자세히 얘기해보시죠 ㅎㅎ 

컴포넌트
> 컴포넌트 분리가 잘 되었는지 궁금합니다!!
> molecules 컴포넌트가 SRP 원칙을 잘 지켰는지 궁금합니다!!


- 코멘트 드렸습니다!

> 컴포넌트간의 의존성이 낮은지, 응집도가 높은지 궁금합니다!!

- HomePage 아래에 notionSidebar, documentDetail 로 나눠서 자식 컴포넌트와 부모 컴포넌트간의 의존성이 높지만 setState -> render 구조 때문에 의존성을 낮추긴 좀 어려워보이긴 하네요 ㅎㅎ 일단은 이정도로 구조로도 괜찮은 거 같습니다.

> styleComponents 처럼 사용하기 위해서 styleInJS라는 utils를 이용해서 스타일을 따로 뺐습니다. 제가 생각했을 때
> 한 파일에서 모든 스타일을 관리하며 classname 중복을 피할 수 있다.
> 컴포넌트 내에 로직부분과 스타일 부분을 따로 분리한다.
> JS문법으로 사용할 수 있다.
> 이러한 특징이 있다고 생각하는데 이게 장점인지 단점인지 헷갈립니다.

- className의 중복을 허용하여 재사용을 높이는데 장점이 있어서 이게 단점이라고 보기는 어렵겠네요 ㅎㅎ 하지만 로직과 스타일을 따로 분리하는 이점, JS로 컨트롤 할 수 있는 이점에 대해서는 동의합니다 ㅎㅎ 모든 기술 선택에는 트레이드오프가 있기 때문에 현재 프로젝트 혹은 자신에게 맞는 이점을 고려하여 선택하면 크게 문제 없다고 생각합니다.

hoilzz commented on src/components/template/NotionSideBar.js:
```diff
@@ -0,0 +1,64 @@
+    const $sidebar = e.target.closest('div[data-notionSideBar]');
+    if (!$sidebar) return;
+
+    const { type, id } = e.target.dataset;
+    if (type === 'document') {
+      e.preventDefault();
+      id && push(`/documents/${id}`);
+    } else {
+      const details = e.target.closest('details');
+      const { id } = details.querySelector('summary span').dataset;
+      if (details.open) {
+        const nextOpenIds = getItem('openDocumentIds').filter(openId => openId !== id);
+        details.setAttribute('open', 'false');
+        setItem('openDocumentIds', nextOpenIds);
+      } else {
+        const currentOpenIds = getItem('openDocumentIds');
+        const nextOpenIds = currentOpenIds ? [...currentOpenIds, id] : [id];
+        setItem('openDocumentIds', nextOpenIds);
+      }
+    }
+  });
```
[이벤트 델리게이션](https://github.com/prgrms-fe-devcourse/FEDC5-5_Project_Notion_VanillaJS/pull/1#discussion_r1376979084) 관련해서 코멘트 참고 부탁드려요 ㅎㅎ

jae-hun-e committed:
[refactor] App.js , api.js, saveDocumentTitle.js 네이밍 수정

App.js
1. route 함수명 => handleRoute
2. 초기설정들 init으로 모아 놓기

api.js
1. api파라미터명 변경

saveDocumentTitle.js
1. 커스텀 이벤트 핸들러 등록 함수와 디스패처 이벤트 함수명 변경
2. 파라미터 함수처럼 보이게 변경

jae-hun-e committed:
[refactor] App.js 다이어트

1. findDocumentDataById 분리

jae-hun-e committed:
[refactor] App.js 다이어트 + NotionTitle.js 이벤트 델리게이션 분해

1. App에서 findDocumentDataById 추출

2. 이벤트는 DOM 내에서만 적용되도록 분해

jae-hun-e committed:
[refactor] createDOM 유팅함수화 + button 이벤트 델리게이션으로 바꾸기

1. DOM 생성부분 유틸 함수화
- DOM생성, 스타일, attribute, appendChild 기능 함수화

2. DocumentObject.js 내의 NewDocumentButton.js, DeleteDocumentButton.js 버튼 이벤트 델리게이션으로 변경

3. NewDocumentButton.js, DeleteDocumentButton.js 내의 hidden 요소 삭제

jae-hun-e committed:
[refactor] molecules SRP 원칙에 맞게 수정

1. DocumentLinkButton.js 안에 있는 로직을 바깥으로 빼서 유틸함수화 (checkCurrentDocument.js)

2. 각 버튼내에 있는 로직, state 바깥으로 뺌(SRP 지키기 위해서 )

jae-hun-e committed:
[refactor] debounce 함수화

1. debounce 함수는 재사용할 일이 많을 것 같아서 유틸함수화함

jae-hun-e committed:
[refactor] DocumentHeader.js => organisms으로 변경

1. DocumentHeader.js는 내부에 molecules인 DocumentLinkButton로 리스트를 만들어 보여주기 때문에 molecules보다 organisms에 어울린다고 생각해서 변경
